File: typescript/2023/day2.ts
import { drop, isNil, max, product, sum } from 'ramda';

type Cube = { color: string; num: number };
type Group = Cube[];
type Game = { gameNum: number; groups: Group[] };

const content = (await Bun.file('2023/inputs/Day2/input.txt').text()).split('\n').filter(Boolean);

// console.log(content);

const parseGame = (str: string): Game => {
  const [gameName, groupsStr] = str.split(': ');

  const gameNum = Number.parseInt(drop(5, gameName), 10);

  const groups = groupsStr.split('; ').map(group =>
    group.split(', ').map(s => {
      const [num, color] = s.split(' ');
      return { color, num: Number.parseInt(num, 10) };
    }),
  );

  return { gameNum, groups };
};

const games = content.map(parseGame);

// part 1
const isGroupPossible = (toCheckAgainst: Group) => (group: Group) =>
  group.every(cube => {
    const found = toCheckAgainst.find(({ color }) => color === cube.color);
    if (isNil(found)) return false;
    return cube.num <= found.num;
  });

const toCheckAgainst: Group = [
  { color: 'red', num: 12 },
  { color: 'green', num: 13 },
  { color: 'blue', num: 14 },
];

const possibleGameNums = games
  .map(({ gameNum, groups }: Game) => ({
    gameNum,
    isPossible: groups.every(isGroupPossible(toCheckAgainst)),
  }))
  .filter(x => x.isPossible)
  .map(({ gameNum }) => gameNum);

const r1 = sum(possibleGameNums);

console.log(r1);

// part 2
const calcGamePower = (game: Game) => {
  const mins = game.groups.flat().reduce<Record<string, number>>((acc, cube) => {
    // eslint-disable-next-line no-param-reassign
    acc[cube.color] = max(acc[cube.color] ?? 0, cube.num);
    return acc;
  }, {});

  return product(Object.values(mins));
};

const r2 = sum(games.map(calcGamePower));

console.log(r2);

File: haskell/2023/Day2.hs
module Day2 where

import Data.Bifunctor
import Data.List.Split
import Data.Map (Map)
import Data.Map qualified as M
import Data.Maybe

type Cube = (String, Int)

type Group = [Cube]

type Game = [Group]

-- parsing
parseCube :: String -> Cube
parseCube s = (color, read num)
  where
    [num, color] = splitOn " " s

parseGroup :: String -> Group
parseGroup = map parseCube . splitOn ", "

parseGame :: String -> (Int, Game)
parseGame s = (gameNum, groups)
  where
    [gameName, gameGroups] = splitOn ": " s
    gameNum = read $ drop 5 gameName
    groups = map parseGroup $ splitOn "; " gameGroups

-- part 1

isGroupPossible :: Group -> Group -> Bool
isGroupPossible toCheckAgainst = all checkCubes
  where
    checkCubes (color, num) = maybe False (num <=) (lookup color toCheckAgainst)

isGamePossible :: Group -> Game -> Bool
isGamePossible toCheckAgainst = all (isGroupPossible toCheckAgainst)

-- part 2

calcGamePower :: Game -> Int
calcGamePower = product . M.elems . foldr (uncurry $ M.insertWith max) M.empty . concat

main' :: IO ()
main' = do
  contents <- lines <$> readFile "../inputs/2023/Day2/input.txt"
  let games = map parseGame contents
  -- part 1
  let toCheckAgainst = [("red", 12), ("green", 13), ("blue", 14)]
  print $ sum $ map fst $ filter snd $ map (second (isGamePossible toCheckAgainst)) games
  -- part 2
  print $ sum $ map (calcGamePower . snd) games
